// devextreme-vue/esm/drawer.js
/*!
 * devextreme-vue
 * Version: 24.1.4
 * Build date: Mon Jul 15 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-vue
 */

import Drawer from "devextreme/ui/drawer";
import { createComponent } from "./core/index";
const DxDrawer = createComponent({
    props: {
        activeStateEnabled: Boolean,
        animationDuration: Number,
        animationEnabled: Boolean,
        closeOnOutsideClick: [Boolean, Function],
        disabled: Boolean,
        elementAttr: Object,
        height: [Function, Number, String],
        hint: String,
        hoverStateEnabled: Boolean,
        maxSize: Number,
        minSize: Number,
        onDisposing: Function,
        onInitialized: Function,
        onOptionChanged: Function,
        opened: Boolean,
        openedStateMode: String,
        position: String,
        revealMode: String,
        rtlEnabled: Boolean,
        shading: Boolean,
        template: {},
        visible: Boolean,
        width: [Function, Number, String]
    },
    emits: {
        "update:isActive": null,
        "update:hoveredElement": null,
        "update:activeStateEnabled": null,
        "update:animationDuration": null,
        "update:animationEnabled": null,
        "update:closeOnOutsideClick": null,
        "update:disabled": null,
        "update:elementAttr": null,
        "update:height": null,
        "update:hint": null,
        "update:hoverStateEnabled": null,
        "update:maxSize": null,
        "update:minSize": null,
        "update:onDisposing": null,
        "update:onInitialized": null,
        "update:onOptionChanged": null,
        "update:opened": null,
        "update:openedStateMode": null,
        "update:position": null,
        "update:revealMode": null,
        "update:rtlEnabled": null,
        "update:shading": null,
        "update:template": null,
        "update:visible": null,
        "update:width": null,
    },
    computed: {
        instance() {
            return this.$_instance;
        }
    },
    beforeCreate() {
        this.$_WidgetClass = Drawer;
        this.$_hasAsyncTemplate = true;
    }
});
export default DxDrawer;
export { DxDrawer };
